test-data/unit/check-python312.test

```diff
@@ -213,7 +213,7 @@ b: Invariant[int]
 if int():
     a = b  # E: Incompatible types in assignment (expression has type "Invariant[int]", variable has type "Invariant[object]")
 if int():
-    b = a  # E: Incompatible types in assignment (expression has type "Invariant[object]", variable has type "Invariant[int]")
+    b = a
```

Comment:
I think this is not right. It happens because with current implementation we check return type after `bind_self()`, so we can't distinguish situation with an explicit `Self` from situation where return type just happens to (accidentally) match current class. I understand that `Self`-types are inherently (slightly) unsafe, but I don't want this unsafety to "leak" to code that doesn't use `Self`.

I think it makes sense to limit the return type erasure only to situations where the original callable type contained an explicit `Self` type (or the "old style" self type).

Reply:
Btw when looking at this I found a bug in variance inference that affects protocols as well. In particular, all three examples above should be inferred as covariant (or even bivariant, but we don't have that).

Also to clarify, the unsafety I meant above relates to examples like
```python
class C[T]:
    def meth(self, x: T, y: C[T]) -> C[T]: ...
```
that should _not_ be contravariant, while it seems to me with the proposed PR it will be. But after playing with this more, inference doesn't behave as it should on master either. This needs some cleanup. @JukkaL we can discuss this in our meeting on Tuesday.